dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+
+    internal ActivityContext(
+        ActivityInfo info,
+        CancellationToken cancellationToken,
+        Action<object[]> heartbeater) { /*...*/ }
+
+    public ActivityInfo Info { get; }
+
+    public CancellationToken CancellationToken { get; }
+
+    public CancellationToken WorkerShutdownToken { get; }
+
+    internal Action<object[]> Heartbeater { get; }
+
+    public void Heartbeat(params object[] details) { /*...*/ }
+}
+```
+
+Notes:
+
+* The current activity context is on thread local or async local depending on whether activity is async or not
+* All calls except `HasCurrent` fail if not in activity context
+* `ActivityInfo` is an immutable record containing info
+* `CancellationToken` and `WorkerShutdownToken` are usable by sync and async
+  * `WorkerShutdownToken` is set on worker shutdown before cancellation token is set. If there is no graceful shutdown
+    period, one is set immediately after the other
+* Note the lack of logger here. See "Logging" later, but in general .NET doesn't have a logger in the standard library.
+  We may make some kind of `IReadOnlyDictionary<string, string> ActivityInfo.LogInfo { get; }` that can be used with
+  `ILogger.BeginScope` or similar.
+
+### Async Completion
+
+Simply `throw new Temporal.Activity.CompleteAsyncException` from the activity.
+
+## Client
+
+Inside the `Temporal.Client` namespace, we have the following types:
+
+* `TemporalClient` - namespace-specific, stateless high-level client to Temporal
+  * See later for most of its API design
+  * Accepts `ITemporalConnection`
+  * Implements `ITemporalClient` for DI use (which itself implements `Temporal.Worker.IWorkerClient` for worker use)
+* `TemporalConnection` - namespace-agnostic, stateful low-level client to Temporal
+  * Has `WorkflowService`, `OperatorService`, etc for making raw gRPC calls
+  * Supports lazy connectivity
+  * Implements `ITemporalConnection` for DI use
+* `WorkflowHandle` - high-level workflow-specific calls
+* Many other supporting types
+
+Example of starting the workflow defined earlier in document, signalling it, then waiting on its results:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Start workflow
+var handle = await client.StartWorkflowAsync(
+    IGreetingWorkflow.Ref.RunAsync, "Temporal", new(ID: "my-workflow-id", TaskQueue: "my-task-queue"));
+
+// Send signal
+await handle.Signal(IGreetingWorkflow.Ref.CompleteWithGreetingAsync);
+
+// Wait for result
+Console.WriteLine(await handle.GetResultAsync());
+```
+
+Example of doing the same via already started workflow:
+
+```csharp
+using Temporal.Client;
+using MyNamespace;
+
+// Connect
+var client = await TemporalClient.ConnectAsync(new("localhost:7233"));
+
+// Get workflow handle
+var handle = await client.GetWorkflowHandle<string>("my-workflow-id");
+
+// Send signal
+await handle.Signal(IGreetingWorkflow.Ref.CompleteWithGreetingAsync);
```

Comment:
Maybe uprade this to the one that sends an arg just so we see what that looks like

Reply:
From the signature later in `WorkflowHandle`, it would be `await handle.Signal(IGreetingWorkflow.Ref.UpdateInfoAsync, new GreetingInfo() { Name = "John" });`. If it is unclear enough I can alter the sample.